### ISSUE TITLE
Security issue when scanning for other webmin servers on the network : password visible in miniserv.log

### ISSUE THREAD

**Issue by litebito** [webmin/webmin#893]
Hi,

when trying to scan for other webmin servers in the network (using the SCAN option, not the BROADCAST option)
I noticed that the credentials are in cleartext in the webserver log of webmin (miniserv.log)

That seems a significant security issue to me? How can that be avoided?

"GET /servers/find.cgi?scan=10.10.10.0&defuser=XXXXX&defpass=YYYYYY&port=10000&undefined= HTTP/1.1" 200 3152

Thanks,

**Comment by litebito** [comment]
That is what I did here, no?

**Comment by qooob** [comment]
Yes, it's Webmin repo already! :dizzy: 

It doesn't 'happen', because it has frames. Most likely, it sends things the same way.

Right click on the link and open it in the new tab, then you will see no left menu and the URL will be displayed. (talking about Gray Theme).

**Comment by gnadelwartz** [comment]
Nevertheless passwords should bever send by get request because url parameters will br logged by webservers in contrast to post parameters...

**Comment by jcameron** [comment]
I'm confused here, because that form is definitely using POST request in the HTML generated by default (at least in the older theme).

**Comment by litebito** [comment]
Maybe the Gray theme sends things the same way, in frames.
but when I look in the logs, I don't see the request with the password when using the gray theme, but I do see it when using the Authentic theme.
And that is my initial main concern : not any log file should contain passwords in clear text
It does not only violate PCIxxx compliance, for example, that system would (should) never pass a security audit in a government environment. It is also against the law in certain countries.

**Comment by qooob** [comment]
In case we have form action set to _POST_, there is no way theme will use something else.

In case there is no action, it's possible that theme guesses it wrong. Make sure that the action is actually set!

**Comment by qooob** [comment]
I meant to say method _POST_.

Yes, it's missing on this page. It has to be sec explicitly to _POST_.

Because it seems that _GET_ is default. [Reference.](https://stackoverflow.com/a/2314441/1455661)

**Comment by qooob** [comment]
Duh, I missed it on `ui_buttons_row`. Sorry about that.

The [patch](https://github.com/qooob/authentic-theme/commit/ab749d3e24599e0e07e9b3ad66ab1f6466066909) will fix it and be going to upcoming release 19.13.

**Comment by gnadelwartz** [comment]
May be we should add a comment to the funtion description to warn theme authors not to create a get form...

**Comment by chris001** [comment]
Can the theme framework  force (or default to) `method=post` forms.

**Comment by qooob** [comment]
@gnadelwartz It's just UI function, that supposed to follow the main-stream logic.

@chris001 UI functions that use forms at the moment are: `ui_form_start` and `ui_buttons_row`. I made `ui_buttons_row` to always be forced to method _POST_, which actually was  on _WebminCore_, which I happen to miss.

It's been fixed.

**Comment by qooob** [comment]
@gnadelwartz There are a lot of places where we use _GET_ method for submitting forms.

**Comment by jcameron** [comment]
Changing the default to POST for all calls to ui_form_start isn't a bad idea though. It shouldn't break anything that I can think of ..

**Comment by chris001** [comment]
Exactly.  The forms should work the same when the `method=post` or `method=get`.  The perl CGI server ought merge the `GET` and `POST` params into `request`. So the framework ought to be able to safely force/default `method=post` for security.

**Comment by jcameron** [comment]
The only case I can see this failing is there's some code in Webmin that doesn't call the `ReadParse` function for parsing CGI inputs, and instead looks at QUERY_STRING directly.

**Comment by qooob** [comment]
@jcameron I will give it a try a bit later. I have technical difficulties right now.

Theme always [runs](https://github.com/qooob/authentic-theme/blob/master/authentic-lib.pm#L3105-L3109) `init_type()`

**Comment by qooob** [comment]
@jcameron @chris001 Method _GET_ is being used in Webmin UI lib, example: 

https://github.com/webmin/webmin/blob/master/ui-lib.pl#L666

**Comment by chris001** [comment]
```
$rv .= "<form class='ui_form' action='".&html_escape($script)."' ".
	($method eq "post" ? "method='post'" :
	 $method eq "form-data" ?
		"method='post' enctype='multipart/form-data'" :
		"method='get'").
	($target ? " target='$target'" : "").
        ($tags ? " ".$tags : "").">\n";
return $rv;
```
It looks like `get` is the default value, when method is not `post` or `form-data`.
Webmin would probably be more secure, and continue to pass tests, if you were to make `post` be the default method, and allow `get` only when specified by the `$method` variable.  In other words, swap the position of `get` and `post` in the code logic flow above.

**Comment by qooob** [comment]
Theoretically this should be working. I already tested it. It's hard to test in every single part of Webmin to tell for sure.

**Comment by chris001** [comment]
Agreed.  Automated free tests with https://travis-ci.org are our only hope for testing every single part of Webmin to tell for sure.

**Comment by qooob** [comment]
@chris001 I'm sorry, I know you mentioned many times about it.

Is this time consuming to deploy it?

**Comment by chris001** [comment]
It's not time consuming to deploy Travis, this is why it's used by one million free open source software projects!
1. Go to https://travis.org
2. Login with github ID/pass
3. Add to Travis `webmin/webmin`, `virtualmin/virtualmin-gpl`, `virtualmin/Virtualmin Install` and `qooob/authentic-theme`.  
4. Enable setting to enable Travis to test all branches of each repo.
5. Enable setting to enable Travis to display test result status in github Pull Requests - this helps repo contributors and repo owners to see if a PR is passing or failing tests.
6. Create new file `.travis.yml` file in the base directory of each repo.

**Comment by qooob** [comment]
@chris001 This is greatly detailed. Thanks a lot. I'm storing your message for future reference.

I will talk with the team about Travis CI.

**Comment by swelljoe** [comment]
We already use Travis CI in places where it is easy to do (new code, or code that has seen some refactoring in recent years). It isn't easy to add to code that doesn't currently have tests that comfortably work in a Travis CI environment. It gets added as we get time.

Volunteers wanting to write unit tests are welcome.

**Comment by chris001** [comment]
Functional tests (for the installers, exit code pass/fail) and `perl` static code analysis (`Perl-critic`), and `codecov`, should really be enabled today, because they're ready to run right now, as they don't require any volunteer to write unit test code.

**Comment by qooob** [comment]
@jcameron We're now defaulting to _POST_ method.

**Comment by qooob** [comment]
@jcameron Like expected, we need _GET_ method used by default, for _forms_ that omit method to be set, to get URL detection in some cases.

This is not security issue and will use the logic, that has been there in WebminCore UI lib for good.

[Reverted.](https://github.com/qooob/authentic-theme/commit/25f482e4ecd6f8529b85cd962212a09bca6281a0)

**Comment by qooob** [comment]
This is also W3C compliant as default method set to _GET_. There is no reason to mess with this around.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
